dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -161,56 +163,107 @@ void RtpsUdpDataLink::do_remove_sample(const RepoId& pub_id,
   g.release();
 
   if (writer) {
-    writer->do_remove_sample(criteria);
+    return writer->remove_sample(sample);
   }
+  return REMOVE_NOT_FOUND;
 }
 
-void
-RtpsUdpDataLink::RtpsWriter::do_remove_sample(const TransportQueueElement::MatchCriteria& criteria)
+void RtpsUdpDataLink::remove_all_msgs(const RepoId& pub_id)
 {
-  SnToTqeMap sn_tqe_map;
-  SnToTqeMap to_deliver;
+  ACE_Guard<ACE_Thread_Mutex> g(lock_);
+  RtpsWriter_rch writer;
+  RtpsWriterMap::iterator iter = writers_.find(pub_id);
+  if (iter != writers_.end()) {
+    writer = iter->second;
+  }
 
-  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+  g.release();
+
+  if (writer) {
+    writer->remove_all_msgs();
+  }
+}
+
+RemoveResult
+RtpsUdpDataLink::RtpsWriter::remove_sample(const DataSampleElement* sample)
+{
+  bool found = false;
+  SequenceNumber to_release;
+  TransportQueueElement* tqe = 0;
+
+  const char* const payload = sample->get_sample()->cont()->rd_ptr();
+  const TransportQueueElement::MatchOnDataPayload modp(payload);
+
+  ACE_Guard<ACE_Thread_Mutex> g(mutex_);
 
   RtpsUdpDataLink_rch link = link_.lock();
   if (!link) {
-    return;
+    return REMOVE_NOT_FOUND;
   }
 
-  ACE_GUARD(ACE_Thread_Mutex, g2, elems_not_acked_mutex_);
+  RemoveResult result = link->send_strategy()->remove_sample(sample);
+
+  ACE_Guard<ACE_Thread_Mutex> g2(elems_not_acked_mutex_);
 
   if (!elems_not_acked_.empty()) {
-    to_deliver.insert(to_deliver_.begin(), to_deliver_.end());
-    to_deliver_.clear();
-    OPENDDS_SET(SequenceNumber) sns_to_release;
     SnToTqeMap::iterator it = elems_not_acked_.begin();
-    while (it != elems_not_acked_.end()) {
-      if (criteria.matches(*it->second)) {
-        sn_tqe_map.insert(RtpsWriter::SnToTqeMap::value_type(it->first, it->second));
-        sns_to_release.insert(it->first);
-        SnToTqeMap::iterator last = it;
-        ++it;
-        elems_not_acked_.erase(last);
+    while (!found && it != elems_not_acked_.end()) {
+      if (modp.matches(*it->second)) {
+        found = true;
+        to_release = it->first;
+        tqe = it->second;
+        elems_not_acked_.erase(it);
       } else {
         ++it;
       }
     }
-    OPENDDS_SET(SequenceNumber)::iterator sns_it = sns_to_release.begin();
-    while (sns_it != sns_to_release.end()) {
-      send_buff_->release_acked(*sns_it);
-      ++sns_it;
-    }
   }
 
   g2.release();
+
+  if (found) {
+    send_buff_->release_acked(to_release);
+  }
+
   g.release();
 
-  SnToTqeMap::iterator deliver_iter = to_deliver.begin();
-  while (deliver_iter != to_deliver.end()) {
-    deliver_iter->second->data_delivered();
-    ++deliver_iter;
+  if (found) {
+    tqe->data_dropped(true);
+    result = REMOVE_FOUND;
   }
+  return result;
+}
+
+void
+RtpsUdpDataLink::RtpsWriter::remove_all_msgs()
+{
+  SnToTqeMap sn_tqe_map;
```

Comment:
Is there a need for this to be default-constructed up front?  Could it just be copied from `elems_not_acked_` on 255?

Reply:
The swap on 255 is intended to be a fast copy + clear of elems_not_acked_ and this is the recipient. I think we still need a valid (default constructed?) destination for the swap to work, unless there's a swap constructor or something I don't know about (move constructors?)

Reply:
That's fine to keep a swap, those are useful.  I guess my comment is just to move the declaration down to where it's needed.